### DISCUSSION TITLE
SQL Server: declaring variable before CTE

### DISCUSSION THREAD

**Discussion by serban-dobroiu** [dbeaver/dbeaver#13854] [Q&A] [answered]
#### System information: 
- Operating system (distribution) and version: Win 10 Version 1809
- DBeaver version: Version 21.1.0.202106012023
- Additional extensions: -

#### Connection specification:
- Database name and version: Microsoft SQL Server Enterprise (64-bit) 11.0.7001.0
- Driver name: Microsoft JDBC Driver for SQL Server
- Do you use tunnels or proxies (SSH, SOCKS, etc)? 

#### Describe the problem you're observing:
This query runs in SSMS but fails with error in DBeaver:
```
DECLARE @test AS INT = 42;
WITH x AS 
	(
	SELECT @test AS 'Column'
	) 
SELECT *
FROM x
```
The error message is 
```
SQL Error [137] [S0002]: Must declare the scalar variable "@test".
```

Removing the semi-colon will expectedly throw:
```
SQL Error [319] [S0001]: Incorrect syntax near the keyword 'with'. If this statement is a common table expression, an xmlnamespaces clause or a change tracking context clause, the previous statement must be terminated with a semicolon.
```
#### Steps to reproduce, if exist:
Run:
```
DECLARE @test AS INT = 42;
WITH x AS 
	(
	SELECT @test AS 'Column'
	) 
SELECT *
FROM x
```

**Comment by uslss** [comment]
thank you for the report

**Comment by serge-rider** [comment, marked as answer]
SQL Server doesn't keep session context between queries.

IF you need to execute several queries which declare/refer some variables then you need to execute them as a single query.

Just select your queries in SQL editor and run them with <kbd>ctrl+enter</kbd>.
